ipa-core/src/protocol/ipa_prf/aggregation/bucket.rs

```diff
@@ -116,26 +115,25 @@ where
         }
 
         let depth_c = ctx.narrow(&BinaryTreeDepthStep::from(tree_depth));
-        let mut futures = Vec::with_capacity(breakdown_count / step);
-
-        for (i, tree_index) in (0..breakdown_count).step_by(step).enumerate() {
-            let bucket_c = depth_c.narrow(&BucketStep::from(i));
-
-            let index_contribution = row_contribution[tree_index].iter();
-
-            if robust || tree_index + span < breakdown_count {
-                futures.push(async move {
-                    let bit_futures = index_contribution
-                        .zip(repeat((bit_of_bdkey, bucket_c.clone())))
-                        .enumerate()
-                        .map(|(i, (a, (b, bucket_c)))| {
-                            a.multiply(b, bucket_c.narrow(&BitOpStep::Bit(i)), record_id)
-                        });
-                    BitDecomposed::try_from(bucket_c.parallel_join(bit_futures).await?)
-                });
-            }
-        }
-        let contributions = ctx.parallel_join(futures).await?;
+
+        let contributions = ctx
+            .parallel_join((0..breakdown_count).step_by(step).enumerate().filter_map(
+                |(i, tree_index)| {
+                    let bucket_c = depth_c.narrow(&BucketStep::from(i));
+
+                    let index_contribution = &row_contribution[tree_index];
+
+                    (robust || tree_index + span < breakdown_count).then(|| {
```

Comment:
I am surprised to see this usage of `.then()`. I thought this was used for chaining two futures together. But this first expression isn't a future, it's just a boolean expression. And we are ignoring the "return value" of this expression (as evidenced by the `||` as the argument to the closure. That seems strange to me. Is this a common usage of `.then()`?

Reply:
It is a [relatively newer method added to `bool`](https://doc.rust-lang.org/std/primitive.bool.html#method.then) as a shorthand for `if /* condition */ { Some(/* value */) } else { None }`. It looks like there are a handful of places we use it, and also a handful where we use the related method `bool::then_some`.

The name collision with the method on futures does seem potentially confusing. I'm not opposed to putting this back the way it was (with a temporary `Vec`), or using the `if / else` longhand rather than `then`, if you think one of those is less confusing.

Reply:
No, this seems like a useful and compact way of expressing things. Thanks for educating me on this! I'm sure I'll wind up using it in the future.